src/Illuminate/Pagination/AbstractCursorPaginator.php

```diff
@@ -0,0 +1,607 @@
+     * @return string
+     */
+    public function url($cursor)
+    {
+        // If we have any extra query string key / value pairs that need to be added
+        // onto the URL, we will put them in query string form and then attach it
+        // to the URL. This allows for extra information like sortings storage.
+        $parameters = is_null($cursor) ? [] : [$this->cursorName => $cursor->encode()];
+
+        if (count($this->query) > 0) {
+            $parameters = array_merge($this->query, $parameters);
+        }
+
+        return $this->path()
+            .(Str::contains($this->path(), '?') ? '&' : '?')
+            .Arr::query($parameters)
+            .$this->buildFragment();
+    }
+
+    /**
+     * Get the URL for the previous page.
+     *
+     * @return string|null
+     */
+    public function previousPageUrl()
+    {
+        if (is_null($previousCursor = $this->previousCursor())) {
+            return null;
+        }
+
+        return $this->url($previousCursor);
+    }
+
+    /**
+     * The URL for the next page, or null.
+     *
+     * @return string|null
+     */
+    public function nextPageUrl()
+    {
+        if (is_null($nextCursor = $this->nextCursor())) {
+            return null;
+        }
+
+        return $this->url($nextCursor);
+    }
+
+    /**
+     * Get the "cursor" of the previous set of items.
+     *
+     * @return \Illuminate\Pagination\Cursor|null
+     */
+    public function previousCursor()
+    {
+        if (is_null($this->cursor) ||
+            ($this->cursor->isPrev() && ! $this->hasMore)) {
+            return null;
+        }
+
+        return $this->getCursorForItem($this->items->first(), false);
+    }
+
+    /**
+     * Get the "cursor" of the next set of items.
+     *
+     * @return \Illuminate\Pagination\Cursor|null
+     */
+    public function nextCursor()
+    {
+        if ((is_null($this->cursor) && ! $this->hasMore) ||
+            (! is_null($this->cursor) && $this->cursor->isNext() && ! $this->hasMore)) {
+            return null;
+        }
+
+        return $this->getCursorForItem($this->items->last(), true);
+    }
+
+    /**
+     * Get a cursor instance for the given item.
+     *
+     * @param  \ArrayAccess|\stdClass  $item
+     * @param  bool  $isNext
+     * @return \Illuminate\Pagination\Cursor
+     */
+    public function getCursorForItem($item, $isNext = true)
+    {
+        return new Cursor($this->getParametersForItem($item), $isNext);
+    }
+
+    /**
+     * @param  \ArrayAccess|\stdClass  $item
+     */
+    public function getParametersForItem($item)
+    {
+        return collect($this->parameters)
+            ->flip()
+            ->map(function ($_, $parameterName) use ($item) {
+                if ($item instanceof ArrayAccess || is_array($item)) {
+                    return $item[$parameterName] ?? $item[Str::afterLast($parameterName, '.')];
+                } elseif ($item instanceof stdClass) {
```

Comment:
How about other objects?

Reply:
The key thing about cursor pagination is that it needs to look inside the paginated items for the parameters to construct the cursor. Since, Eloquent and query builders support `ArrayAccess` and `stdClass`, both are supported here. 

If you have any other object in mind that you'd like to support, would be happy to amend this.

Reply:
How about this? I think the extra flexibility will not hurt.

```php
if (is_object($item)) {
  return $item->$parameterName ...
```

Reply:
This is very helpful. I've included this, thanks 👍